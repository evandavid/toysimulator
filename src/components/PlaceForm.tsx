/* eslint-disable react-native/no-inline-styles */
/* eslint-disable radix */
import React, {useEffect, useState} from 'react';
import {Modal, TouchableWithoutFeedback, View, Text} from 'react-native';
import {ExecuteButton, ButtonText, Row} from '../styled';
import {ROBOT_POSITION} from '../util';
import Radio from './Radio';
import {
  PlaceFormContainer,
  PlaceFormOuterContainer,
  Overlay,
  Input,
  Label,
} from './styled';

type PlaceFormProps = {
  visible: boolean;
  onRequestClose: () => void;
  onSubmit: (data: ROBOT_POSITION) => void;
};

const PlaceForm = ({visible, onRequestClose, onSubmit}: PlaceFormProps) => {
  const [form, setForm] = useState<ROBOT_POSITION>({
    col: 0,
    row: 0,
    f: 'SOUTH',
  });

  const onExecute = () => {
    onSubmit(form);
  };

  useEffect(() => {
    if (!visible) {
      /** reset form */
      setForm({
        col: 0,
        row: 0,
        f: 'SOUTH',
      });
    }
  }, [visible]);

  return (
    <Modal transparent visible={visible} onRequestClose={onRequestClose}>
      <PlaceFormOuterContainer>
        <TouchableWithoutFeedback onPress={onRequestClose}>
          <Overlay />
        </TouchableWithoutFeedback>

        <PlaceFormContainer>
          <View>
            <Label>X (Col)</Label>
            <Input
              placeholder="X value"
              keyboardType="number-pad"
              value={String(form.col)}
              onChangeText={value => {
                const numberValue = parseInt(value);
                if (isNaN(numberValue)) {
                  setForm({...form, col: 0});
                  return;
                }
                if (numberValue < 5) {
                  setForm({...form, col: numberValue});
                }
              }}
            />
          </View>
          <View>
            <Label>Y (Row)</Label>
            <Input
              placeholder="Y value"
              keyboardType="number-pad"
              value={String(form.row)}
              onChangeText={value => {
                const numberValue = parseInt(value);
                if (isNaN(numberValue)) {
                  setForm({...form, row: 0});
                  return;
                }
                if (numberValue < 5) {
                  setForm({...form, row: numberValue});
                }
              }}
            />
          </View>
          <View style={{margin: 12, marginBottom: 0}}>
            <Row>
              <TouchableWithoutFeedback
                onPress={() => {
                  setForm({...form, f: 'NORTH'});
                }}>
                <Row style={{flex: 1, marginTop: 6}}>
                  <Radio selected={form.f === 'NORTH'} />
                  <Text style={{color: '#212121'}}>NORTH</Text>
                </Row>
              </TouchableWithoutFeedback>
              <TouchableWithoutFeedback
                onPress={() => {
                  setForm({...form, f: 'EAST'});
                }}>
                <Row style={{flex: 1, marginTop: 6}}>
                  <Radio selected={form.f === 'EAST'} />
                  <Text style={{color: '#212121'}}>EAST</Text>
                </Row>
              </TouchableWithoutFeedback>
            </Row>
            <Row>
              <TouchableWithoutFeedback
                onPress={() => {
                  setForm({...form, f: 'WEST'});
                }}>
                <Row style={{flex: 1, marginTop: 6}}>
                  <Radio selected={form.f === 'WEST'} />
                  <Text style={{color: '#212121'}}>WEST</Text>
                </Row>
              </TouchableWithoutFeedback>
              <TouchableWithoutFeedback
                onPress={() => {
                  setForm({...form, f: 'SOUTH'});
                }}>
                <Row style={{flex: 1, marginTop: 6}}>
                  <Radio selected={form.f === 'SOUTH'} />
                  <Text style={{color: '#212121'}}>SOUTH</Text>
                </Row>
              </TouchableWithoutFeedback>
            </Row>
          </View>
          <View
            style={{
              height: 36,
              marginTop: 12,
              marginLeft: 6,
              marginRight: 6,
            }}>
            <TouchableWithoutFeedback onPress={onExecute}>
              <ExecuteButton>
                <ButtonText>SAVE</ButtonText>
              </ExecuteButton>
            </TouchableWithoutFeedback>
          </View>
          <View
            style={{
              height: 36,
              marginLeft: 6,
              marginRight: 6,
            }}>
            <TouchableWithoutFeedback onPress={onRequestClose}>
              <ExecuteButton style={{backgroundColor: '#3D5866'}}>
                <ButtonText>CANCEL</ButtonText>
              </ExecuteButton>
            </TouchableWithoutFeedback>
          </View>
        </PlaceFormContainer>
      </PlaceFormOuterContainer>
    </Modal>
  );
};

export default PlaceForm;
