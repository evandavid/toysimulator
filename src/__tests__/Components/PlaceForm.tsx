import React from 'react';
import renderer from 'react-test-renderer';
import PlaceForm from '../../components/PlaceForm';
import {render, fireEvent} from '@testing-library/react-native';

beforeEach(() => {
  jest.useFakeTimers();
});

it('renders correctly', () => {
  const tree = renderer
    .create(
      <PlaceForm visible={false} onRequestClose={jest.fn} onSubmit={jest.fn} />,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();

  const tree1 = renderer
    .create(
      <PlaceForm visible={true} onRequestClose={jest.fn} onSubmit={jest.fn} />,
    )
    .toJSON();
  expect(tree1).toMatchSnapshot();
});

it('handle onSubmit correctly', () => {
  const onPressSubmitMock = jest.fn();
  const {getByText} = render(
    <PlaceForm
      visible={false}
      onRequestClose={jest.fn}
      onSubmit={onPressSubmitMock}
    />,
  );

  fireEvent.press(getByText('SAVE'));
  expect(onPressSubmitMock).toHaveBeenCalled();

  // return default value
  expect(onPressSubmitMock).toHaveBeenCalledWith({col: 0, row: 0, f: 'SOUTH'});
});

it('handle X input change correctly', () => {
  const onPressSubmitMock = jest.fn();
  const {getByText, getByPlaceholderText} = render(
    <PlaceForm
      visible={false}
      onRequestClose={jest.fn}
      onSubmit={onPressSubmitMock}
    />,
  );
  fireEvent.changeText(getByPlaceholderText('X value'), 1);
  fireEvent.press(getByText('SAVE'));
  expect(onPressSubmitMock).toHaveBeenCalled();
  expect(onPressSubmitMock).toHaveBeenCalledWith({col: 1, row: 0, f: 'SOUTH'});
});

it('handle Y input change correctly', () => {
  const onPressSubmitMock = jest.fn();
  const {getByText, getByPlaceholderText} = render(
    <PlaceForm
      visible={false}
      onRequestClose={jest.fn}
      onSubmit={onPressSubmitMock}
    />,
  );
  fireEvent.changeText(getByPlaceholderText('Y value'), 1);
  fireEvent.press(getByText('SAVE'));
  expect(onPressSubmitMock).toHaveBeenCalled();
  expect(onPressSubmitMock).toHaveBeenCalledWith({col: 0, row: 1, f: 'SOUTH'});
});

it('handle F  change correctly', () => {
  const onPressSubmitMock = jest.fn();
  const {getByText} = render(
    <PlaceForm
      visible={false}
      onRequestClose={jest.fn}
      onSubmit={onPressSubmitMock}
    />,
  );
  fireEvent.press(getByText('EAST'));
  fireEvent.press(getByText('SAVE'));
  expect(onPressSubmitMock).toHaveBeenCalled();
  expect(onPressSubmitMock).toHaveBeenCalledWith({col: 0, row: 0, f: 'EAST'});
});
