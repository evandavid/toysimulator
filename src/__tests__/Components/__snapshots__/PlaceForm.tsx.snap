// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Modal
  hardwareAccelerated={false}
  onRequestClose={[Function]}
  transparent={true}
  visible={false}
>
  <View
    className="sc-crrszt fWleXe"
  >
    <View
      accessible={true}
      className="sc-bqyKOL johKCT"
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
    />
    <View
      className="sc-dQoVA eoKiZr"
    >
      <View>
        <Text
          className="sc-hBEYId jvKEGi"
        >
          X (Col)
        </Text>
        <TextInput
          className="sc-kstqJO QsSFc"
          keyboardType="number-pad"
          onChangeText={[Function]}
          placeholder="X value"
          value="0"
        />
      </View>
      <View>
        <Text
          className="sc-hBEYId jvKEGi"
        >
          Y (Row)
        </Text>
        <TextInput
          className="sc-kstqJO QsSFc"
          keyboardType="number-pad"
          onChangeText={[Function]}
          placeholder="Y value"
          value="0"
        />
      </View>
      <View
        style={
          Object {
            "margin": 12,
            "marginBottom": 0,
          }
        }
      >
        <View
          className="sc-gsTEea jwzyjw"
        >
          <View
            accessible={true}
            className="sc-gsTEea jwzyjw"
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "flex": 1,
                "marginTop": 6,
              }
            }
          >
            <View
              className="sc-fodVek ceasiC"
            />
            <Text
              style={
                Object {
                  "color": "#212121",
                }
              }
            >
              NORTH
            </Text>
          </View>
          <View
            accessible={true}
            className="sc-gsTEea jwzyjw"
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "flex": 1,
                "marginTop": 6,
              }
            }
          >
            <View
              className="sc-fodVek ceasiC"
            />
            <Text
              style={
                Object {
                  "color": "#212121",
                }
              }
            >
              EAST
            </Text>
          </View>
        </View>
        <View
          className="sc-gsTEea jwzyjw"
        >
          <View
            accessible={true}
            className="sc-gsTEea jwzyjw"
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "flex": 1,
                "marginTop": 6,
              }
            }
          >
            <View
              className="sc-fodVek ceasiC"
            />
            <Text
              style={
                Object {
                  "color": "#212121",
                }
              }
            >
              WEST
            </Text>
          </View>
          <View
            accessible={true}
            className="sc-gsTEea jwzyjw"
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "flex": 1,
                "marginTop": 6,
              }
            }
          >
            <View
              className="sc-fodVek ceasiC"
            >
              <View
                className="sc-fFucqa cxhKwi"
              />
            </View>
            <Text
              style={
                Object {
                  "color": "#212121",
                }
              }
            >
              SOUTH
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "height": 36,
            "marginLeft": 6,
            "marginRight": 6,
            "marginTop": 12,
          }
        }
      >
        <View
          accessible={true}
          className="sc-gKseQn sc-iBPTik bEKjdd lhQWQt"
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
        >
          <Text
            className="sc-fubCzh GEpea"
          >
            SAVE
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "height": 36,
            "marginLeft": 6,
            "marginRight": 6,
          }
        }
      >
        <View
          accessible={true}
          className="sc-gKseQn sc-iBPTik bEKjdd lhQWQt"
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "backgroundColor": "#3D5866",
            }
          }
        >
          <Text
            className="sc-fubCzh GEpea"
          >
            CANCEL
          </Text>
        </View>
      </View>
    </View>
  </View>
</Modal>
`;

exports[`renders correctly 2`] = `
<Modal
  hardwareAccelerated={false}
  onRequestClose={[Function]}
  transparent={true}
  visible={true}
>
  <View
    className="sc-crrszt fWleXe"
  >
    <View
      accessible={true}
      className="sc-bqyKOL johKCT"
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
    />
    <View
      className="sc-dQoVA eoKiZr"
    >
      <View>
        <Text
          className="sc-hBEYId jvKEGi"
        >
          X (Col)
        </Text>
        <TextInput
          className="sc-kstqJO QsSFc"
          keyboardType="number-pad"
          onChangeText={[Function]}
          placeholder="X value"
          value="0"
        />
      </View>
      <View>
        <Text
          className="sc-hBEYId jvKEGi"
        >
          Y (Row)
        </Text>
        <TextInput
          className="sc-kstqJO QsSFc"
          keyboardType="number-pad"
          onChangeText={[Function]}
          placeholder="Y value"
          value="0"
        />
      </View>
      <View
        style={
          Object {
            "margin": 12,
            "marginBottom": 0,
          }
        }
      >
        <View
          className="sc-gsTEea jwzyjw"
        >
          <View
            accessible={true}
            className="sc-gsTEea jwzyjw"
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "flex": 1,
                "marginTop": 6,
              }
            }
          >
            <View
              className="sc-fodVek ceasiC"
            />
            <Text
              style={
                Object {
                  "color": "#212121",
                }
              }
            >
              NORTH
            </Text>
          </View>
          <View
            accessible={true}
            className="sc-gsTEea jwzyjw"
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "flex": 1,
                "marginTop": 6,
              }
            }
          >
            <View
              className="sc-fodVek ceasiC"
            />
            <Text
              style={
                Object {
                  "color": "#212121",
                }
              }
            >
              EAST
            </Text>
          </View>
        </View>
        <View
          className="sc-gsTEea jwzyjw"
        >
          <View
            accessible={true}
            className="sc-gsTEea jwzyjw"
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "flex": 1,
                "marginTop": 6,
              }
            }
          >
            <View
              className="sc-fodVek ceasiC"
            />
            <Text
              style={
                Object {
                  "color": "#212121",
                }
              }
            >
              WEST
            </Text>
          </View>
          <View
            accessible={true}
            className="sc-gsTEea jwzyjw"
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "flex": 1,
                "marginTop": 6,
              }
            }
          >
            <View
              className="sc-fodVek ceasiC"
            >
              <View
                className="sc-fFucqa cxhKwi"
              />
            </View>
            <Text
              style={
                Object {
                  "color": "#212121",
                }
              }
            >
              SOUTH
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "height": 36,
            "marginLeft": 6,
            "marginRight": 6,
            "marginTop": 12,
          }
        }
      >
        <View
          accessible={true}
          className="sc-gKseQn sc-iBPTik bEKjdd lhQWQt"
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
        >
          <Text
            className="sc-fubCzh GEpea"
          >
            SAVE
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "height": 36,
            "marginLeft": 6,
            "marginRight": 6,
          }
        }
      >
        <View
          accessible={true}
          className="sc-gKseQn sc-iBPTik bEKjdd lhQWQt"
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "backgroundColor": "#3D5866",
            }
          }
        >
          <Text
            className="sc-fubCzh GEpea"
          >
            CANCEL
          </Text>
        </View>
      </View>
    </View>
  </View>
</Modal>
`;
